File: apps/api/src/core/demo_clock.py
"""Opt-in wall clock for the private local demo environment.

Production always uses the real UTC clock. The override is intentionally
guarded by ``ENVIRONMENT=local`` and has no default; the tech-demo Compose
profile is the only shipped deployment that sets it.
"""
from __future__ import annotations

import os
from datetime import datetime, timezone


def utcnow() -> datetime:
    raw = os.getenv("SHUTTLEWORKS_DEMO_NOW")
    if not raw:
        return datetime.now(timezone.utc)
    if os.getenv("ENVIRONMENT", "local").strip().lower() != "local":
        raise RuntimeError("SHUTTLEWORKS_DEMO_NOW is allowed only when ENVIRONMENT=local")
    try:
        parsed = datetime.fromisoformat(raw.replace("Z", "+00:00"))
    except ValueError as exc:
        raise RuntimeError("SHUTTLEWORKS_DEMO_NOW must be an ISO-8601 datetime") from exc
    if parsed.tzinfo is None:
        raise RuntimeError("SHUTTLEWORKS_DEMO_NOW must include a UTC offset")
    return parsed.astimezone(timezone.utc)


__all__ = ["utcnow"]

File: apps/console/src/lib/demoClock.ts
/** Stable clock for the private demo; unset builds use the real clock. */
function parseDemoNow(raw: string | undefined, environment: string | undefined): number | null {
  if (!raw) return null;
  if ((environment ?? 'production').trim().toLowerCase() !== 'local') {
    throw new Error('VITE_DEMO_NOW is allowed only when VITE_ENVIRONMENT=local');
  }
  const parsed = Date.parse(raw);
  if (!Number.isFinite(parsed) || !/[zZ]|[+-]\d{2}:?\d{2}$/.test(raw.trim())) {
    throw new Error('VITE_DEMO_NOW must be an ISO-8601 datetime with a UTC offset');
  }
  return parsed;
}

export function demoNowMs(fallbackMs: number = Date.now()): number {
  return parseDemoNow(import.meta.env.VITE_DEMO_NOW, import.meta.env.VITE_ENVIRONMENT) ?? fallbackMs;
}

export function demoNow(fallback?: Date): Date { return new Date(demoNowMs(fallback?.getTime())); }

export { parseDemoNow };
